### DISCUSSION TITLE
Is the admin_rule table cleaned up after extension uninstalling?

### DISCUSSION THREAD

**Discussion by func0der** [OpenMage/magento-lts#3381] [Questions & Answers] [answered]
Debugging another problem, I noticed today, that my `exception.log`  file is spammed with messages like this:

```
Zend_Acl_Exception: Resource 'admin/system/config/ayalinemostviewed' not found in /path/to/openmage/vendor/shardj/zf1-future/library/Zend/Acl.php:365
Stack trace:
#0 /path/to/openmage/vendor/shardj/zf1-future/library/Zend/Acl.php(641): Zend_Acl->get()
#1 /path/to/openmage/vendor/shardj/zf1-future/library/Zend/Acl.php(519): Zend_Acl->setRule()
#2 /path/to/openmage/webroot/app/code/core/Mage/Admin/Model/Resource/Acl.php(133): Zend_Acl->deny()
#3 /path/to/openmage/webroot/app/code/core/Mage/Admin/Model/Resource/Acl.php(71): Mage_Admin_Model_Resource_Acl->loadRules()
#4 /path/to/openmage/webroot/app/code/core/Mage/Admin/Model/Session.php(163): Mage_Admin_Model_Resource_Acl->loadAcl()
#5 /path/to/openmage/webroot/app/code/core/Mage/Admin/Model/Observer.php(63): Mage_Admin_Model_Session->login()
#6 /path/to/openmage/webroot/app/code/core/Mage/Core/Model/App.php(1399): Mage_Admin_Model_Observer->actionPreDispatchAdmin()
#7 /path/to/openmage/webroot/app/code/core/Mage/Core/Model/App.php(1377): Mage_Core_Model_App->_callObserverMethod()
#8 /path/to/openmage/webroot/app/Mage.php(508): Mage_Core_Model_App->dispatchEvent()
#9 /path/to/openmage/webroot/app/code/core/Mage/Core/Controller/Varien/Action.php(528): Mage::dispatchEvent()
#10 /path/to/openmage/webroot/app/code/core/Mage/Adminhtml/Controller/Action.php(174): Mage_Core_Controller_Varien_Action->preDispatch()
#11 /path/to/openmage/webroot/app/code/core/Mage/Core/Controller/Varien/Action.php(411): Mage_Adminhtml_Controller_Action->preDispatch()
#12 /path/to/openmage/webroot/app/code/core/Mage/Core/Controller/Varien/Router/Standard.php(256): Mage_Core_Controller_Varien_Action->dispatch()
#13 /path/to/openmage/webroot/app/code/core/Mage/Core/Controller/Varien/Front.php(182): Mage_Core_Controller_Varien_Router_Standard->match()
#14 /path/to/openmage/webroot/app/code/core/Mage/Core/Model/App.php(365): Mage_Core_Controller_Varien_Front->dispatch()
#15 /path/to/openmage/webroot/app/Mage.php(742): Mage_Core_Model_App->run()
#16 /path/to/openmage/webroot/index.php(55): Mage::run()
#17 {main}
````

This module that is mentioned in the error message I have uninstalled a long time ago and by 'uninstall' I mean, I just stopped deploying it with the rest of the application.
I do not know if that is the right way to do it, but I know of none other.

Is there a mechanism in place to clean those leftover entries in `admin_rule`?
Does anything clean up after 'uninstalled' modules?

**Comment by addison74** [comment, marked as answer]
We discussed a similar issue here https://github.com/OpenMage/magento-lts/issues/3175.

These are the **admin_role**, **admin_rule** and **admin_user** tables from an (OpenMage + Magento Sample Data) installation. You could check their content, and if everything is fine, look for the problem elsewhere.

![admin_role](https://github.com/OpenMage/magento-lts/assets/8360474/057cfe90-ab1a-43fc-8c10-e9350579b1c4)
![admin_rule](https://github.com/OpenMage/magento-lts/assets/8360474/49913d3c-7865-4edf-a64a-3a3137776da2)
![admin_user](https://github.com/OpenMage/magento-lts/assets/8360474/3dd8b309-60dc-48fa-b7e3-342b25478394)

As far as uninstalling an extension is concerned, it is a complex process. If the developer has provided instructions or a script, it's fine, but I found that everything is not always deleted and orphaned files and records remain in the database.

I have my own rule to analyze an extension in a test environment before using it. The first time I analyze the files and see where they are copied, then I look in the files that add records to the database. I move on to installing the extension and running it again, then I compare the database before and after installation (sometimes I need to compare the file system when uploading files). That's the only way I have control over all the data and can proceed to a complete uninstallation later.

Answering to your question, there is no "mechanism" to clean up orphan records from the database (there is an extension to clean orphan attribute values). You have to do the whole process manually, only after analyzing the extension as I showed above.

**Comment by func0der** [comment]
Somehow a weird cache problem seems to have caused these errors: https://github.com/OpenMage/magento-lts/issues/3383

Now those errors seem to be gone.

But it is good to know, that I have to cleanup myself after modules (most of the time) :D

**Reply by addison74** [reply]
Production extensions/modules are rarely uninstalled, because they are established in the development process. When problems occur, they can be temporarily disabled until the installation is checked in a test environment. The file system is the easiest to clean, attention should be directed to the database (if the extension modifies the database).
